src/main/java/org/broadinstitute/hellbender/tools/dataflow/MarkDuplicatesDataflow.java

```diff
@@ -124,37 +170,42 @@ public int partitionFor(final GATKRead read, final int n) {
-         *   (a) The remaining reads (one per fragment key) are coordinate-sorted and paired
-         *       consecutively together and emitted.
-         *   (b) If a read is leftover, it is emitted, unmodified, as an unpaired end.
-         * (3) Paired ends are grouped by a similar key as above but using both reads.
-         * (4) Any unpaired end is emitted, unmodified.
-         * (5) The remained paired ends are scored and all but the highest scoring are marked as
-         *     duplicates. Both reads in the pair are emitted.
+         * (1) labelReadsByName: label each read with its read group and read name.
+         * (2) GroupByKey: group together reads with the same group and name.
+         * (3) labelPairedEndsWithAlignmentInfo:
+         *   (a) Sort each group of reads (see LexicographicSort below).
+         *   (b) Pair consecutive reads into PairedEnds. In most cases there will only be two reads
+         *       with the same name. TODO: explain why there might be more.
+         *   (c) Label each read with alignment information: Library, reference index,
+         *       stranded unclipped start and reverse strand.
+         *   (d) Leftover reads are emitted, unmodified, as an unpaired end.
+         * (4) GroupByKey: Group PairedEnds that share alignment information. These pairs
+         *     are duplicates of each other.
+         * (5) markDuplicatePairs:
+         *   (a) For each group created by (4), sort the pairs by score and mark all but the
+         *       highest scoring as duplicates.
+         *   (b) Determine which duplicates are optical duplicates and increase the overall count.
          */
         private PCollection<GATKRead> transformReads(final PCollectionView<SAMFileHeader> headerPcolView, final PCollection<GATKRead> pairs) {
-            final PTransform<PCollection<? extends GATKRead>, PCollection<KV<String, GATKRead>>> makeKeysForPairs = makeKeysForPairs(headerPcolView);
-            final PTransform<PCollection<? extends KV<String, Iterable<GATKRead>>>, PCollection<KV<String, PairedEnds>>> markGroupedDuplicatePairs = markGroupedDuplicatePairs(headerPcolView);
-            final PTransform<PCollection<? extends KV<String, Iterable<PairedEnds>>>, PCollection<GATKRead>> markPairedEnds = markPairedEnds();
+            final PTransform<PCollection<? extends GATKRead>, PCollection<KV<String, GATKRead>>> labelReadsByName = labelReadsByName(headerPcolView);
+            final PTransform<PCollection<? extends KV<String, Iterable<GATKRead>>>, PCollection<KV<String, PairedEnds>>> labelPairedEndsWithAlignmentInfo =
+                labelPairedEndsWithAlignmentInfo(headerPcolView);
+            final PTransform<PCollection<? extends KV<String, Iterable<PairedEnds>>>, PCollection<GATKRead>> markDuplicatePairedEnds = markDuplicatePairedEnds();
 
             return pairs
-                    .apply(makeKeysForPairs)
+                    .apply(labelReadsByName)
```

Comment:
I think including key in the name is useful to help mentally map back to what's going on in dataflow.   Maybe `keyReadsByName`?